#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [79]:
import pandas as pd
df = pd.read_csv('course/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [19]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [47]:
grouped = df[['Branch','gross income']].groupby('Branch')
grouped.sum().sort_values('gross income', ascending = False)

gross income
Branch              
C          5265.1765
A          5057.1605
B          5057.0320

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [148]:
grouped2 = df[['City', 'gross income']].groupby('City')
grouped2.sum().sort_values('gross income', ascending = False).head(1)

gross income
City                   
Naypyitaw     5265.1765

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [160]:
invoices_by_category = pd.pivot_table(df, index = 'Branch', columns = 'Product line', values = 'Invoice ID', aggfunc = "count")
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [209]:
invoices_by_category.sort_values('Electronic accessories', ascending = False).head(1)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59

опція 2, вивести через max якимось простішим способом не вдалось

In [229]:
max_invoices = max(invoices_by_category["Electronic accessories"])
invoices_by_category.query('`Electronic accessories` == @max_invoices')

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. Хто робить більше покупок - чоловіки чи жінки? 
2. Які найпопулярніші групи товарів серед чоловіків та жінок за кількістю покупок (invoice id)? 

In [308]:
popular_products = df[df.Gender == "Female"].groupby('Product line')
popular_products[ "Invoice ID"].count()


Product line
Electronic accessories    84
Fashion accessories       96
Food and beverages        90
Health and beauty         64
Home and lifestyle        79
Sports and travel         88
Name: Invoice ID, dtype: int64

In [353]:
female_filter = df[df.Gender == "Female"]
male_filter = df[df.Gender == "Male"]
products_female = female_filter["Product line"].value_counts().sort_values(ascending = False)
products_male = male_filter["Product line"].value_counts().sort_values(ascending = False)
products_female, sum(products_female), products_male, sum(products_male)

(Product line
 Fashion accessories       96
 Food and beverages        90
 Sports and travel         88
 Electronic accessories    84
 Home and lifestyle        79
 Health and beauty         64
 Name: count, dtype: int64,
 501,
 Product line
 Health and beauty         88
 Electronic accessories    86
 Food and beverages        84
 Fashion accessories       82
 Home and lifestyle        81
 Sports and travel         78
 Name: count, dtype: int64,
 499)

**Висновок**: бачимо, що різниця в кількості покупок чоловіків та жінок - мінімальна (499 vs 501). Тому тут не маємо конкретних рекомендації, змін, адже наші покупці збалансовані по гендеру, і нам треба орієнтуватись на обидва гендери. 
Щодо категорії товарів, бачимо що у жінок найпопулярніші Fashion accessories та Food and begerages; тоді як в чоловіків -  Health and beauty та Electronic accessories. Таким чином, знаючи ці том 4 категорії, ми можемо підсилити їх маркетинг, переконатись що наявні товари достатні в кількості та асортименті для того щоб підвищити ще більше продажі по цим категоріям. 
Також бачимо що  Home and lifestyle передостання категорія для обох гендерів. Можливо варто дослідити з чии це може бути повязано та зрозуміти можливості для ціієї категорії.

3. Хто робить більше покупок Member чи Normal? Хто робить покупки на більші суми? 

In [336]:
df['Customer type'].value_counts()

Customer type
Member    501
Normal    499
Name: count, dtype: int64

In [342]:
pd.pivot_table(df, index = ['Customer type'], values = ['Total'], aggfunc = ('count', 'sum', 'mean'))

Total                        
              count        mean         sum
Customer type                              
Member          501  327.791305  164223.444
Normal          499  318.122856  158743.305

**Висновок**: бачимо, що попри майже однакову кількість покупок (501 VS 499), Member мають вищу середню суму покупки (327 VS 318) та в результаті більшу загальну суму усіх покупок. Для подальшого розвитку бізнесу, можна переглянути наявну програму для того щоб стати Member - в першу чергу, можна розглянути наскільки легкий та швидкий цей процес. Наприклад, в деяких магазинах це вимагає заповнення анкети на касі, чи повідомлення особистих данних, тоді як в інших - є альтернативна можливість просто завантажити додаток, що може бути кращою опцією для молодшої аудиторії. 